File: s1/s1/src/main/java/com/nodal/s1/consumer/RabbitMQConsumer.java
package com.nodal.s1.consumer;

import com.nodal.s1.publisher.RabbitMQPublisher;
import org.apache.tomcat.jni.Time;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class RabbitMQConsumer {

    @Value("${rabbitmq.routing.key.s1.response}")
    private String respondS1RoutingKey;

    @Value("${rabbitmq.exchange.name}")
    private String exchange;

    private static final Logger LOGGER = LoggerFactory.getLogger(RabbitMQConsumer.class);

    private final RabbitTemplate rabbitTemplate;

    public RabbitMQConsumer(RabbitTemplate rabbitTemplate) {
        this.rabbitTemplate = rabbitTemplate;
    }
    @Autowired
    private RabbitMQPublisher rabbitMQPublisher;

    @RabbitListener(queues = {"${rabbitmq.queue.s1.send}"})
    public void consumeSend(String message){
        //publishPong after receiving Ping from s2

        LOGGER.info("S1 received following message from S2: {}", message);
        if("Ping".equals(message)) {
            rabbitTemplate.convertAndSend(exchange, respondS1RoutingKey, "Pong");
            LOGGER.info("S1 sends response with message: Pong");
            //wait 10 seconds
            try {
                Thread.sleep(10000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            LOGGER.info("S1 sending 'Ping' to S2");
            rabbitMQPublisher.sendMessage("Ping");
        }
        logSplitter();
    }

    @RabbitListener(queues = {"${rabbitmq.queue.s2.response}"})
    public void consumeResponse(String message){
        //publishPong after receiving Ping from s2
        LOGGER.info("S1 received following Response message S2: {}", message);
        logSplitter();
    }

    private void logSplitter(){
        LOGGER.info("---------------------------------------------------------");
    }
}

File: s2/s2/src/main/kotlin/com/nodal/s2/consumer/RabbitMQConsumer.kt
package com.nodal.s2.consumer

import com.nodal.s2.publisher.RabbitMQPublisher
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
class RabbitMQConsumer {
    @Value("\${rabbitmq.routing.key.s2.response}")
    private lateinit var respondS2RoutingKey: String

    @Value("\${rabbitmq.exchange.name}")
    private lateinit var exchange: String

    val LOGGER: Logger = LoggerFactory.getLogger(RabbitMQConsumer::class.java)

    private var rabbitTemplate: RabbitTemplate

    constructor(rabbitTemplate: RabbitTemplate) {
        this.rabbitTemplate = rabbitTemplate
    }

    @Autowired
    private val rabbitMQPublisher: RabbitMQPublisher? = null

    @RabbitListener(queues = ["\${rabbitmq.queue.s2.send}"])
    fun consume(message: String?) {

        LOGGER.info("S2 received following message from S1: {}", message)
        if("Ping" == message) {
            rabbitTemplate.convertAndSend(exchange, respondS2RoutingKey, "Pong")

            try {
                Thread.sleep(10000)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
            }

            LOGGER.info("S2 sending 'Ping' to S1")
            rabbitMQPublisher!!.sendMessage("Ping")
        }
        logSplitter()
    }

    @RabbitListener(queues = ["\${rabbitmq.queue.s1.response}"])
    fun consumeResponse(message: String?) {
        //publishPong after receiving Ping from s2

        LOGGER.info("S2 received following Response message from S1: {}", message)
        logSplitter()
    }

    private fun logSplitter() {
        LOGGER.info("---------------------------------------------------------")
    }

}